duplex_http_call.go

```diff
@@ -316,3 +339,77 @@ func cloneURL(oldURL *url.URL) *url.URL {
 	}
 	return newURL
 }
+
+// bufferPipeWriter is a write-through cache.
+type bufferPipeWriter struct {
+	*io.PipeWriter
+
+	mu     sync.Mutex
+	buffer *bytes.Buffer
+}
+
+// Write to the PipeWriter and buffer the data if available.
+func (b *bufferPipeWriter) Write(data []byte) (int, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	bytesWritten, err := b.PipeWriter.Write(data)
+	if b.buffer == nil || err != nil {
+		return bytesWritten, err
+	}
+	_, _ = b.buffer.Write(data[:bytesWritten])
```

Comment:
All `buffer := bufferpool.Get(); defer bufferpool.Put(buffer)` would be unsafe. We'd need to redo all bufferpool access to change the liveness scope to be the request and not per send/recv.

I think if we go that route it makes more sense to have one `*bytes.Buffer` that is written to. We also still need to manage long running streams which shouldn't hold onto old buffers forever.

Reply:
Correct that the sender cannot release them back to the pool. Instead, it would be the _reader_ that must release them back to the pool.

I definitely disagree about using a single `*bytes.Buffer`. I don't see how that helps you decipher message boundaries or helps avoid copying.